game/state/city/building.cpp

```diff
@@ -96,6 +97,48 @@ bool Building::hasAliens() const
 	return false;
 }
 
+void Building::initBuilding(GameState &state)
+{
+	// Initialize economy data, done in the map/city editor or when game starts for the first time
+	// Not on save/load, that's why values are serialized
+	currentWage = 65;
+	maximumWorkforce = countActiveTiles() * function->workersPerTile * 50 / 100;
+	currentWorkforce = maximumWorkforce * 70 / 100;
+	maintenanceCosts = randBoundsInclusive(state.rng, 90, 110) * function->baseCost / 100;
+	incomePerCapita = randBoundsInclusive(state.rng, 90, 110) * function->baseIncome / 100;
+	investment = function->investmentValue;
+	prestige = function->prestige;
+}
+
+void Building::updateWorkforce()
+{
+	maximumWorkforce = countActiveTiles() * function->workersPerTile * 50 / 100;
+	if (maximumWorkforce < currentWorkforce)
+	{
+		city->populationUnemployed -= currentWorkforce - maximumWorkforce;
+		currentWorkforce = maximumWorkforce;
+	}
+}
+
+int Building::calculateIncome() const
+{
+	return currentWorkforce * (incomePerCapita - currentWage) - maintenanceCosts;
+}
+
+unsigned Building::countActiveTiles() const
+{
+	unsigned relevantTiles = 0;
+	for (auto &p : buildingParts)
+	{
+		auto tile = city->map->getTile(p);
+		if (tile->presentScenery && tile->presentScenery->type->value > 5)
```

Comment:
Not a big fan of the magic number "5" - if there's a limited set of types it makes sense to have a static set of values, or even modifying the scenerytype to have a flag that says "This is part of a building"

Reply:
What about making it `constexpr`? Or is there a config file we can put this kind of variables into?

Reply:
Stuff like that should probably go in the GameState - as if mods add new scenery types they probably also want to alter this value.

But I think it may be better to add a new bool flag to the SceneryType object itself - and use that to key off. For the initial data files, you can add something to the extractor to set the flag if the type.value > 5 - as we 'know' the OG data files use that value, but any future mods that add new/modify scenery types can set that flag themselves as necessary.

Reply:
I like this approach! Will add the boolean.